pkg/util/deployment/deployment.go

```diff
@@ -409,3 +420,52 @@ func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired
 
 	return surge, unavailable, nil
 }
+
+// used for unit testing
+var nowFn = func() time.Time { return time.Now() }
+
+// TimedOut returns if the provided deployment is progressing or timed out.
+func TimedOut(d extensions.Deployment) bool {
+	if d.Spec.ProgressDeadlineSeconds == nil || len(d.Status.Conditions) == 0 {
+		return false
+	}
+
+	latestCondition := d.Status.Conditions[len(d.Status.Conditions)-1]
```

Comment:
This is also useful for observing the behavior of a deployment thoughout its lifetime.